infrastructure/src/main/java/org/corfudb/infrastructure/log/FileChannelPerf.java

```diff
@@ -0,0 +1,182 @@
+        }
+        update(readMetrics, READ_METRICS, start, size);
+        return res;
+    }
+
+    public int read(ByteBuffer dst) throws IOException {
+        return read(dst, false, 0);
+    }
+
+    public int read(ByteBuffer dst, long position) throws IOException {
+        return read(dst, true, position);
+    }
+
+    public int write(ByteBuffer buf, boolean use_position, long position) throws IOException {
+        long start = System.nanoTime();
+        int size = buf.remaining();
+        int res = 0;
+        if(!use_position)
+            res = fileChannel.write(buf);
+        else {
+            res = fileChannel.write(buf, position);
+        }
+        update(writeMetrics, WRITE_METRICS, start, size);
+        return res;
+    }
+
+    public int write(ByteBuffer dst) throws IOException {
+        return write(dst, false, 0);
+    }
+
+    public int write(ByteBuffer dst, long position) throws IOException {
+        return write(dst, true, position);
+    }
+
+    public void force(boolean force) throws IOException {
+        final long start = System.nanoTime();
+        fileChannel.force(force);
+
+        if(force)
+            update(syncMetaMetrics, SYNC_META_METRICS, start, MB_DATA);
+        else
+            update(syncMetrics, SYNC_METRICS, start, MB_DATA);
+    }
+
+    public static void metricsHis(Timer timer, String name) {
+        log.info("Timer {} Count {} Latency mean: {} ms  50%: {} ms  95%: {} ms  99%: {} ms",
+                timer,
+                timer.getCount(),
+                timer.getSnapshot().getMean() / 1000000,
+                timer.getSnapshot().get75thPercentile() / 1000000,
+                timer.getSnapshot().get95thPercentile() / 1000000,
+                timer.getSnapshot().get99thPercentile() / 1000000);
+    }
+
+    public static void metricsHis() {
+        metricsHis(readMetrics, READ_METRICS);
+        metricsHis(writeMetrics, WRITE_METRICS);
+        metricsHis(syncMetrics, SYNC_METRICS);
+        metricsHis(syncMetaMetrics, SYNC_META_METRICS);
+    }
+
+    public final boolean isOpen() {
+        return fileChannel.isOpen();
+    }
+
+    public FileChannelPerf truncate(final long size) throws IOException {
+        fileChannel.truncate(size);
+        return this;
+    }
+
+    public void close() throws IOException {
+        fileChannel.close();
+    }
+
+    public static synchronized void update(Timer timer, String name, long start, long size) {
+        long dur = (System.nanoTime() - start);
+        long durSec = TimeUnit.SECONDS.convert(System.nanoTime() - start, TimeUnit.NANOSECONDS);
+
+        if (durSec > spikeLatency.getSeconds() * getUnit(size)) {
+            log.warn("There is a spike: {} {} MB data in {} seconds.", name, getUnit(size), durSec);
+        }
+        slidingWindow.update(durSec);
+        timer.update(dur, TimeUnit.NANOSECONDS);
+    }
+
+    public static boolean reportSpike() {
+        return detectorEnabled && slidingWindow.report();
```

Comment:
If latency spikes, but this isn't called and the time window passes by then the calling this won't report a spike, is this what you expect? Once a spike is reported then it should persist, no?

Reply:
The goal is that when the spike observed in a fair long period, it means either the disk, storage system, file system, or jvm has serious problem. If the spike goes away, it means the system comes back healthy again.